third_party/move/move-compiler-v2/src/pipeline/explicit_drop.rs

```diff
@@ -0,0 +1,197 @@
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AssignKind, AttrId, Bytecode, Operation},
+};
+use std::collections::BTreeSet;
+
+pub struct ExplicitDrop {}
+
+impl FunctionTargetProcessor for ExplicitDrop {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        mut data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let target = FunctionTarget::new(fun_env, &data);
+        let mut transformer = ExplicitDropTransformer::new(target);
+        transformer.transform();
+        data.code = transformer.transformed;
+        data.annotations.remove::<LiveVarAnnotation>();
+        data.annotations.remove::<LifetimeAnnotation>();
+        data
+    }
+
+    fn name(&self) -> String {
+        "ExplicitDrop".to_owned()
+    }
+}
+
+struct ExplicitDropTransformer<'a> {
+    target: FunctionTarget<'a>,
+    // result of the transformation
+    transformed: Vec<Bytecode>,
+    live_var_annot: &'a LiveVarAnnotation,
+    lifetime_annot: &'a LifetimeAnnotation,
+}
+
+impl<'a> ExplicitDropTransformer<'a> {
+    pub fn new(target: FunctionTarget<'a>) -> Self {
+        let live_var_annot = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("livevar annotation");
+        let lifetime_annot = target
+            .get_annotations()
+            .get::<LifetimeAnnotation>()
+            .expect("lifetime annotation");
+        ExplicitDropTransformer {
+            target,
+            transformed: Vec::new(),
+            live_var_annot,
+            lifetime_annot,
+        }
+    }
+
+    /// Add explicit drop instructions
+    /// note that this will invalidate existing analyses
+    pub fn transform(&mut self) {
+        self.drop_unused_args();
+        for (code_offset, bytecode) in self.target.get_bytecode().to_vec().iter().enumerate() {
+            self.emit_bytecode(bytecode.clone());
+            self.explicate_drops_at(code_offset as CodeOffset, bytecode);
+        }
+    }
+
+    // add drops at given code offset
+    fn explicate_drops_at(&mut self, code_offset: CodeOffset, bytecode: &Bytecode) {
+        match bytecode {
+            Bytecode::Ret(..) | Bytecode::Jump(..) | Bytecode::Abort(..) | Bytecode::Branch(..) => {
+                ()
+            },
+            _ => {
+                let released_temps = self.released_temps_at(code_offset);
+                self.drop_temps(&released_temps, bytecode.get_attr_id())
+            },
+        }
+    }
+
+    fn drop_unused_args(&mut self) {
+        let code_offset = 0;
+        let live_var_info = self.get_live_var_info(code_offset);
+        let lifetime_info = self.get_lifetime_info(code_offset);
+        for arg in self.target.get_parameters() {
```

Comment:
Why do we not filter out primitives here?

Reply:
It's a response to https://github.com/aptos-labs/aptos-core/pull/11119#discussion_r1423107875.
Just for the sake of ability checking, we don't need drops for primitives. This is also to align with the treatment of primitive types in the copy transformation. We don't copy locals of primitives explicitly in function call arguments even if it's necessary.
Also, we need to make it clear how each instruction reads its arguments: are the sources moved (as in move assignments), copied (as in copy assignments), or referenced (as in borrow_loc's)? Before this is made clear, I cannot argue for the correctness of the explicit drop pass.
But anyway the current implementation suffices for ability checking.

Reply:
Sorry, I am not sure this answers my question. I understand we don't need drops for primitives, which is what you do in `released_temps_at()`. But, why do we also not filter out dropping primitives here (that is, within `drop_unused_args()`)? What is different here that we add explicit drops here even for primitives?

Reply:
Oops, I misread the question and thought you were asking the opposite. Yes, I just forgot to filter out the primitives here.

Reply:
done